sql/connect/server/src/main/scala/org/apache/spark/sql/connect/pipelines/PipelinesHandler.scala

```diff
@@ -0,0 +1,288 @@
+          Option(graphElementRegistry.defaultCatalog),
+          Option(graphElementRegistry.defaultDatabase)),
+        comment = None,
+        origin = QueryOrigin(
+          objectType = Option(QueryOriginType.Flow.toString),
+          objectName = Option(flowIdentifier.unquotedString),
+          language = Option(Python()))))
+  }
+
+  private def startRun(
+      cmd: proto.PipelineCommand.StartRun,
+      responseObserver: StreamObserver[ExecutePlanResponse],
+      sessionHolder: SessionHolder): Unit = {
+    val dataflowGraphId = cmd.getDataflowGraphId
+    val graphElementRegistry = DataflowGraphRegistry.getDataflowGraphOrThrow(dataflowGraphId)
+    // We will use this variable to store the run failure event if it occurs. This will be set
+    // by the event callback.
+    var runFailureEvent = Option.empty[PipelineEvent]
+    // Define a callback which will stream logs back to the SparkConnect client when an internal
+    // pipeline event is emitted during pipeline execution. We choose to pass a callback rather the
+    // responseObserver to the pipelines execution code so that the pipelines module does not need
+    // to take a dependency on SparkConnect.
+    val eventCallback = { event: PipelineEvent =>
+      val message = if (event.error.nonEmpty) {
+        // Returns the message associated with a Throwable and all its causes
+        def getExceptionMessages(throwable: Throwable): Seq[String] = {
+          throwable.getMessage +:
+            Option(throwable.getCause).map(getExceptionMessages).getOrElse(Nil)
+        }
+        val errorMessages = getExceptionMessages(event.error.get)
+        s"""${event.message}
+           |Error: ${errorMessages.mkString("\n")}""".stripMargin
+      } else {
+        event.message
+      }
+      event.details match {
+        // Failed runs are recorded in the event log. We do not pass these to the SparkConnect
+        // client since the failed run will already result in an unhandled exception that is
+        // propagated to the SparkConnect client. This special handling ensures that the client
+        // does not see the same error twice for a failed run.
+        case RunProgress(state) if state == FAILED => runFailureEvent = Some(event)
+        case RunProgress(state) if state == CANCELED =>
+          throw new RuntimeException("Pipeline run was canceled.")
+        case _ =>
+          responseObserver.onNext(
+            proto.ExecutePlanResponse
+              .newBuilder()
+              .setSessionId(sessionHolder.sessionId)
+              .setServerSideSessionId(sessionHolder.serverSessionId)
+              .setPipelineEventResult(
+                proto.PipelineEventResult.newBuilder
+                  .setEvent(
+                    proto.PipelineEvent
+                      .newBuilder()
+                      .setTimestamp(ProtoTimestamp
+                        .newBuilder()
+                        .setSeconds(event.timestamp.getTime / 1000)
```

Comment:
I am not sure if we have to document this, initially it seemed that this is was arbitrarily dropping/adding milliseconds. However java.sql.Timestamp normalizes its time and nanos fields.